cheatsheets/Java_Security_Cheat_Sheet.md

```diff
@@ -0,0 +1,685 @@
+- [Usage of OWASP Security Logging with Logback](https://github.com/javabeanz/owasp-security-logging/wiki/Usage-with-Logback)
+
+## Cryptography
+
+### General cryptography guidance
+
+- **Never, ever write your own cryptographic functions.**
+- Make sure your application or protocol can easily support a future change of cryptographic algorithms.
+- Use your package manager whereever possible to keep all of your packages up to date. Watch the updates on your development setup, and plan updates to your applications accordingly.
+
+### Encryption for storage
+
+Follow the algorithm guidance in the [OWASP Cryptographic Storage Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Cryptographic_Storage_Cheat_Sheet.html#algorithms).
+
+The following code snippet shows an example of using AES-GCM to perform encryption/decryption of data.
+
+```java
+import java.nio.charset.StandardCharsets;
+import java.security.SecureRandom;
+import javax.crypto.spec.*;
+import javax.crypto.*;
+import java.util.Base64;
+
+
+// AesGcmSimpleTest
+class Main {
+
+    public static void main(String[] args) throws Exception {
+        // Key of 32 bytes / 256 bits for AES
+        KeyGenerator keyGen = KeyGenerator.getInstance(AesGcmSimple.ALGORITHM);
+        keyGen.init(AesGcmSimple.KEY_SIZE, new SecureRandom());
+        SecretKey secretKey = keyGen.generateKey();
+
+        // Nonce of 12 bytes / 96 bits and this size should always be used.
+        byte[] nonce = new byte[AesGcmSimple.IV_LENGTH];
+        SecureRandom random = new SecureRandom();
+        random.nextBytes(nonce);
+
+        var message = "This message to be encrypted";
+        System.out.println(message);
+
+        // Encrypt the message
+        byte[] cipherText = AesGcmSimple.Encrypt(message, nonce, secretKey);
+        System.out.println(Base64.getEncoder().encodeToString(cipherText));
+
+        // Decrypt the message
+        var message2 = AesGcmSimple.Decrypt(cipherText, nonce, secretKey);
+        System.out.println(message2);
+    }
+}
+
+class AesGcmSimple {
+
+    public static final String ALGORITHM = "AES";
+    public static final String CIPHER_ALGORITHM = "AES/GCM/NoPadding";
+    public static final int KEY_SIZE = 256;
+    public static final int TAG_LENGTH = 128;
+    public static final int IV_LENGTH = 12;
+
+    public static byte[] Encrypt(String plaintext, byte[] nonce, SecretKey secretKey) throws Exception {
+        return CryptoOperation(plaintext.getBytes(StandardCharsets.UTF_8), nonce, secretKey, Cipher.ENCRYPT_MODE);
+    }
+
+    public static String Decrypt(byte[] ciphertext, byte[] nonce, SecretKey secretKey) throws Exception {
+        return new String(CryptoOperation(ciphertext, nonce, secretKey, Cipher.DECRYPT_MODE), StandardCharsets.UTF_8);
+    }
+
+    private static byte[] CryptoOperation(byte[] text, byte[] nonce, SecretKey secretKey, int mode) throws Exception {

```

Comment:
And where is `CryptoOperation` defined?

Reply:
Not sure I understand your question. I just used this method as otherwise the code from lines 560-563 gets duplicate for encrypt and decrypt.

Reply:
The question was I didn't see a definition for CryptoOperation, so it seems to be missing. It's possible that I missed it though.

Reply:
@kwwall can you confirm what the issue here as I have indeed defined the function...